Return the number of bytes successfully written even if there is an error.

Happy0 opened the pull request:
I'm using this library for a library that is quite write heavy. Sometimes, the `write` call only manages to send *some* of the bytes. 

As well as the error string, I'd like the number of bytes that were *successfully* written to be returned so that my application can retry to send the rest (but not repeat the bytes that *were* successfully written.)

eelcocramer commented:
Thanks for the PR. LGTM. Just as a sanity check. The API is still backward compatible right? As you can pass an arbitrary number of parameters to a function in javascript, it will not break the code of a developer that is still using the _old_ callback.

Happy0 commented:
I think so, @eelcocramer ? I just tried adding a arbitrary number of parameters at the end of the callback, and it seemed to still work:

```
    device.write( msg, (err, bytes, test1, test2) => {
```

I was thinking... Maybe the 'error' value of the callback should only be populated if the write was a complete failure? i.e. the result of the `write` call is `-1`? Partial writes (for example, the `write` call succeeding for 900 bytes out of 1000 written) aren't so unusual in IO, so in my opinion the callback should be ` cb(null, 900) ` in such the case of not all bytes being transferred rather than `cb("Writing attempt was not completely successful", null)`. 

In the case of *no* bytes being sent, I'd expect cb("No bytes written", 0)

However, I think this is less important than exposing the number of bytes that were written on every callback.

What do you think?

Also, I'm thinking we probably shouldn't populate the 'bytes' parameter with -1 (like this pull request currently does) if no bytes were written as a client might use this parameter to calculate how many bytes in the buffer to retry,

eelcocramer commented:
Usually in IO a write will return the number of bytes written. If the returned value is less then the length of the buffer that was supposed to be written the output was truncated. If the return value is negative an error did occur.

Following that paradigm the callback function could be something like `cb(err, size)`. If `size >= 0` than `err` must be a `nil` value. If `size < 0` than `err` must contains the error that occurred.

It would be even better if the first parameter of the callback would be the `size` but this would break the API.

Happy0 commented:
Cheers, @eelcocramer . I've updated the pull request to implement this behaviour.

I've been able to test it compiles and works on Linux, but I haven't been able to do the same for windows / mac (I only have a Linux laptop.)

I tried making the necessary modifications to the windows implementation - but I don't trust myself to go near the mac one. I don't even know what a `.mm` file is or what language that is :P.

eelcocramer commented:
Thanks again for the update!

But I think it is a good idea to set `size` to a negative value in case of an error, like you implemented it before but just not set the `err` object if `size >= 0`. This will `force` devs to implement some kind of error handling but give them the option to check for either the `err` object or the `size < 0`.

eelcocramer commented:
Also, the `.mm` are just C++ with Objective C mixed in. It needs the `.mm` extension to be recognised by the compiler.

Happy0 commented:
@eelcocramer fair play!

Should I update the documentation an examples to change `bytesWritten` (as a variable name / function callback parameter name) to `result`? To me, `bytesWritten` implies that it'd always be >=0 as it doesn't make say to say "-1 bytes were written."

With a name like 'bytesWritten', the application developer might assume it's >=0, and do their retry like

```
btSerial.write( buffer.slice(bytesWritten, buffer.length )
```

For example:

```
var btSerial = new (require('bluetooth-serial-port')).BluetoothSerialPort();

btSerial.on('found', function(address, name) {
	btSerial.findSerialPortChannel(address, function(channel) {
		btSerial.connect(address, channel, function() {
			console.log('connected');

            var buffer = new Buffer('my data', 'utf-8')

			btSerial.write(buffer, function(err, result) {
				if (err || result < 0) {
                    // Retry or fail.
                } else if (result < buffer.length) {
                    // Some bytes were successfully written, retry the rest...
                    var bytesWritten = result;
                    var restOfBuffer = buffer.slice(bytesWritten, buffer.length);

                    btSerial.write(restOfBuffer, ...);
                }
			});

			btSerial.on('data', function(buffer) {
				console.log(buffer.toString('utf-8'));
			});
		}, function () {
			console.log('cannot connect');
		});

		// close the connection when you're ready
		btSerial.close();
	}, function() {
		console.log('found nothing');
	});
});

btSerial.inquire();
```

eelcocramer commented:
I agree that `result` would be a better name.

Happy0 commented:
I've updated the pull request to retain the -1 result, and updated the documentation to reflect that the second parameter is the 'result', not the number of bytes written.

Let me know if you need anything else :)

eelcocramer commented:
Closed. Merged and will be part of the next release (v2.2.0)

eelcocramer commented:
So I've merged this and I was doing a final check for the OSX code and I noticed that I've implemented write to be reliable. Than I checked the windows implementation and that is reliable as well.

So now I'm considering reverting some of this PR and make the Linux implementation reliable as well.

Happy0 commented:
@eelcocramer . sure, that works too :)

eelcocramer commented:
Just opened PR #194 which aligns the linux implementation of write with those of windows and macos. Do you have the chance to test if this fix works for you?
